src/Symfony/Component/Config/Definition/JsonSchemaGenerator.php

```diff
@@ -0,0 +1,190 @@
+    {
+        $schema = array_replace_recursive([
+            '$schema' => 'http://json-schema.org/draft-06/schema#',
+            'definitions' => [
+                'param' => [
+                    '$comment' => 'Container parameter',
+                    'type' => 'string',
+                    'pattern' => '^%[^%]+%$',
+                ],
+                'root' => $this->buildSingleNode($node, allowParam: false),
+            ],
+            '$ref' => '#/definitions/root',
+        ], $schema);
+
+        file_put_contents($this->outputPath, json_encode($schema, \JSON_PRETTY_PRINT | \JSON_UNESCAPED_SLASHES | \JSON_THROW_ON_ERROR));
+    }
+
+    private function buildArrayNode(ArrayNode $node): array
+    {
+        $schema = [
+            'type' => ['object', 'array'],
+            'additionalProperties' => $node->shouldIgnoreExtraKeys(),
+            'maxItems' => 0,
+        ];
+
+        foreach ($node->getChildren() as $child) {
+            $schema['properties'][$child->getName()] = $this->buildSingleNode($child);
+        }
+
+        return $schema;
+    }
+
+    private function buildSingleNode(NodeInterface $node, bool $allowParam = true): array|\ArrayObject
+    {
+        $schema = match (\count($types = $this->createSubSchemas($node, $allowParam))) {
+            1 => $types[0],
+            default => ['anyOf' => $types],
+        };
+
+        if ($node->hasDefaultValue()) {
+            $schema['default'] = $node->getDefaultValue();
+        }
+
+        if ($node instanceof BaseNode) {
+            if ($info = $node->getInfo()) {
+                $schema['description'] = $info;
+            }
+
+            if ($node->isDeprecated()) {
+                $schema['deprecated'] = true;
+            }
+        }
+
+        return $schema;
+    }
+
+    private function createSubSchemas(NodeInterface $node, bool $allowParam = true): array
+    {
+        $paramTypes = [];
+
+        $getType = fn ($value) => match (get_debug_type($value)) {
+            'string' => 'string',
+            'int' => 'integer',
+            default => null,
+        };
+
+        $removeNulls = fn (array $array) => array_filter($array, fn ($value) => null !== $value);
+
+        if ($node instanceof BaseNode && !$node instanceof StringNode && \in_array(ExprBuilder::TYPE_STRING, $node->getNormalizedTypes(), true)) {
+            $paramTypes[] = ['type' => 'string'];
+        }
+
+        $pseudoType = match (true) {
+            $node instanceof BooleanNode => 'bool',
+            $node instanceof IntegerNode => 'int',
+            $node instanceof NumericNode => 'float',
+            $node instanceof StringNode => 'string',
+            $node instanceof EnumNode => 'enum',
+            $node instanceof PrototypedArrayNode => 'array_prototype',
+            $node instanceof ArrayNode => 'array',
+            $node instanceof ScalarNode => 'scalar',
+            default => null,
```

Comment:
I don't think there are any cases that don't fit into the above. I would delete the default to leave an error if the class is not expected.

Reply:
As I understand, it's possible to have a completely custom `NodeInterface` implementation in user land, and if you did that, I think it'd be better to allow anything in the subschema than to fail generating the whole schema?